framework/wazuh/core/indexer/__init__.py

```diff
@@ -0,0 +1,120 @@
+import os
+import random
+from asyncio import sleep
+from logging import getLogger
+
+from opensearchpy import AsyncOpenSearch
+from wazuh.core.exception import WazuhIndexerError
+
+logger = getLogger('wazuh')
+
+HOST_KEY = 'host'
+PORT_KEY = 'port'
+
+
+class Indexer:
+    """Interface to connect with Wazuh Indexer."""
+
+    def __init__(
+        self,
+        host: str,
+        user: str,
+        password: str,
+        port: int = 9200,
+        use_ssl: bool = True,
+        verify_certs: bool = True,
+        ca_certs: str = '',
+    ) -> None:
+        self.host = host
+        self.user = user
+        self.password = password
+        self.port = port
+        self.use_ssl = use_ssl
+        self.verify_certs = verify_certs
+        self.ca_certs = ca_certs
+
+        self._client = self._get_opensearch_client()
+
+        # Register index clients here
+
+    def _get_opensearch_client(self) -> AsyncOpenSearch:
+        """Get the a new instance of the opensearch client.
+
+        Returns
+        -------
+        AsyncOpenSearch
+            The created instance.
+        """
+        return AsyncOpenSearch(
+            hosts=[{HOST_KEY: self.host, PORT_KEY: self.port}],
+            http_compress=True,
+            http_auth=(self.user, self.password),
+            use_ssl=self.use_ssl,
+            verify_certs=self.verify_certs,
+            ca_certs=self.ca_certs,
+        )
+
+    async def initialize(self):
+        """Initialize the Wazuh Indexer connection.
+
+        Raises
+        ------
+        WazuhIndexerError
+            In case of errors communicating with the Wazuh Indexer.
+        """
+        if not (await self._client.ping()):
+            raise WazuhIndexerError(2200)
+
+    async def close(self):
+        """Close the Wazuh Indexer client."""
+        await self._client.close()
+
+
+async def create_indexer(
+    host: str,
+    user: str,
+    password: str,
+    port: int = 9200,
+    retries: int = 5,
+    backoff_in_seconds: int = 1,
+    **kwargs,
+) -> Indexer:
+    """Create and initialize the Indexer instance implementing a retry with backoff mechanism.
+
+    Parameters
+    ----------
+    host : str
+        Location of the Wazuh Indexer.
+    user : str
+        User of the Wazuh Indexer to authenticate with.
+    password : str
+        Password of the Wazuh Indexer to authenticate with.
+    port : int, optional
+        Port of the Wazuh Indexer to connect with, by default 9200
+    retries : int, optional
+        Number of retries, by default 5.
+    backoff_in_seconds : int, optional
+        Base seconds to wait, by default 1.
+
+    Returns
+    -------
+    Indexer
+        The new Indexer instance.
+    """
+
+    indexer = Indexer(host, user, password, port, **kwargs)
```

Comment:
Is the connection opened here? Because `initialize()` just pings the endpoint to see if there's any response, but if the server is down it will probably fail before reaching that call. The solution may be including this instantiation inside the try/catch.

Reply:
This is not needed. The connection pool is populated after the first request.

```python
ipdb> n
> /home/nstefani/git/wazuh/framework/wazuh/core/indexer/__init__.py(65)connect()
     64         """
---> 65         print('Connection pool: ', self._client.transport.connection_pool)
     66         if not (await self._client.ping()):

ipdb> n
Connection pool:  <EmptyConnectionPool: []>
> /home/nstefani/git/wazuh/framework/wazuh/core/indexer/__init__.py(66)connect()
     65         print('Connection pool: ', self._client.transport.connection_pool)
---> 66         if not (await self._client.ping()):
     67             raise WazuhIndexerError(2200)

ipdb> n
> /home/nstefani/git/wazuh/framework/wazuh/core/indexer/__init__.py(68)connect()
     67             raise WazuhIndexerError(2200)
---> 68         print('Connection pool: ', self._client.transport.connection_pool)
     69

ipdb> n
Connection pool:  <DummyConnectionPool: (<AIOHttpConnection: http://127.0.0.1:9200>,)>
```